x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/DatafeedCcsIT.java

```diff
@@ -0,0 +1,214 @@
+            .put(XPackSettings.SECURITY_ENABLED.getKey(), false)
+            .put(XPackSettings.WATCHER_ENABLED.getKey(), false)
+            .put(XPackSettings.GRAPH_ENABLED.getKey(), false)
+            .put(LifecycleSettings.LIFECYCLE_HISTORY_INDEX_ENABLED_SETTING.getKey(), false)
+            // Default the watermarks to absurdly low to prevent the tests from failing on nodes without enough disk space
+            .put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING.getKey(), "1b")
+            .put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_WATERMARK_SETTING.getKey(), "1b")
+            .put(DiskThresholdSettings.CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING.getKey(), "1b")
+            .build();
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins(String clusterAlias) {
+        return List.of(
+            LocalStateMachineLearning.class,
+            CommonAnalysisPlugin.class,
+            IngestCommonPlugin.class,
+            ReindexPlugin.class,
+            ShutdownPlugin.class,
+            // To remove warnings about painless not being supported
+            BaseMlIntegTestCase.MockPainlessScriptEngine.TestPlugin.class,
+            // ILM is required for .ml-state template index settings
+            IndexLifecycle.class,
+            // Deprecation warnings go to a data stream, if we ever cause a deprecation warning the data streams plugin is required
+            DataStreamsPlugin.class,
+            // To remove errors from parsing built in templates that contain scaled_float
+            MapperExtrasPlugin.class
+        );
+    }
+
+    @Override
+    protected Collection<String> remoteClusterAlias() {
+        return List.of(REMOTE_CLUSTER);
+    }
+
+    @Override
+    protected boolean reuseClusters() {
+        return false;
+    }
+
+    public void testDatafeedWithCcsRemoteHealthy() throws Exception {
+        setSkipUnavailable(randomBoolean());
+        String jobId = "ccs-healthy-job";
+        String datafeedId = jobId;
+        long numDocs = randomIntBetween(32, 2048);
+        long endTimeMs = indexRemoteDocs(numDocs);
+        setupJobAndDatafeed(jobId, datafeedId, endTimeMs);
+        // Datafeed should complete and auto-close the job
+        assertBusy(() -> {
+            GetJobsStatsAction.Response statsResponse = client(LOCAL_CLUSTER).execute(
+                GetJobsStatsAction.INSTANCE,
+                new GetJobsStatsAction.Request(jobId)
+            ).actionGet();
+            assertThat(statsResponse.getResponse().results().get(0).getState(), is(JobState.CLOSED));
+            assertThat(statsResponse.getResponse().results().get(0).getDataCounts().getProcessedRecordCount(), is(numDocs));
+        }, 30, TimeUnit.SECONDS);
+        clearSkipUnavailable();
+    }
+
+    public void testDatafeedWithCcsRemoteUnavailable() throws Exception {
```

Comment:
This test often passes, and proves the code changes work.

However, the PR cannot be merged as-is because sometimes this test fails as follows:

```
	at __randomizedtesting.SeedInfo.seed([41E6F5EF31368549:15BE9EBA233BDCEA]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.test.InternalTestCluster.lambda$assertSearchContextsReleased$44(InternalTestCluster.java:2529)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1068)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1041)
	at org.elasticsearch.test.InternalTestCluster.assertSearchContextsReleased(InternalTestCluster.java:2528)
	at org.elasticsearch.test.InternalTestCluster.assertAfterTest(InternalTestCluster.java:2470)
	at org.elasticsearch.test.AbstractMultiClustersTestCase.assertAfterTest(AbstractMultiClustersTestCase.java:139)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

The post-test assertions sometimes find unreleased search contexts.  So there must be somewhere in the datafeed code that doesn't release search contexts if a search fails.

In addition there are a couple of outstanding TODOs in this test, but I thought I'd push the code in its current form for visibility.

Reply:
I think the issue is that we need to clear scroll if the search response contains skipped clusters.